File: app/routers/scores.py
import json
from datetime import date
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlite3 import Connection
from typing import List, Optional

from app.database import get_db
from app.models import ScoreAddRequest, EvaluationRuleCreate, UndoRequest
from app.timezone import get_today_str_taipei, get_now_str_taipei
from app.ws_manager import manager

router = APIRouter(prefix="/api/scores", tags=["Scores"])

@router.get("/{course_id}/rules")
def list_rules(course_id: int, db: Connection = Depends(get_db)):
    cursor = db.cursor()
    cursor.execute("""
        SELECT * FROM evaluation_rules WHERE course_id = ? ORDER BY category DESC, id ASC
    """, (course_id,))
    return [dict(row) for row in cursor.fetchall()]

@router.post("/{course_id}/rules")
def create_rule(course_id: int, data: EvaluationRuleCreate, db: Connection = Depends(get_db)):
    cursor = db.cursor()
    cursor.execute("""
        INSERT INTO evaluation_rules (course_id, title, category, score_value, icon)
        VALUES (?, ?, ?, ?, ?)
    """, (course_id, data.title, data.category, data.score_value, data.icon))
    db.commit()
    return {"id": cursor.lastrowid, "message": "Rule created"}

DEFAULT_RULES = [
    {"title": "熱心助人", "category": "positive", "score_value": 1, "icon": "🤝"},
    {"title": "發言踴躍", "category": "positive", "score_value": 1, "icon": "🙋‍♂️"},
    {"title": "專心聽講", "category": "positive", "score_value": 1, "icon": "👂"},
    {"title": "作業優良", "category": "positive", "score_value": 2, "icon": "📝"},
    {"title": "團隊合作", "category": "positive", "score_value": 1, "icon": "🌟"},
    {"title": "上課吵鬧", "category": "negative", "score_value": -1, "icon": "📢"},
    {"title": "未帶用品", "category": "negative", "score_value": -1, "icon": "🎒"},
    {"title": "上課分心", "category": "negative", "score_value": -1, "icon": "😴"},
]

@router.put("/{course_id}/rules/{rule_id}")
def update_rule(course_id: int, rule_id: int, data: EvaluationRuleCreate, db: Connection = Depends(get_db)):
    cursor = db.cursor()
    cursor.execute("""
        UPDATE evaluation_rules
        SET title = ?, category = ?, score_value = ?, icon = ?
        WHERE id = ? AND course_id = ?
    """, (data.title, data.category, data.score_value, data.icon, rule_id, course_id))
    db.commit()
    return {"message": "Rule updated"}

@router.post("/{course_id}/rules/reset_defaults")
def reset_default_rules(course_id: int, db: Connection = Depends(get_db)):
    cursor = db.cursor()
    cursor.execute("DELETE FROM evaluation_rules WHERE course_id = ?", (course_id,))
    for rule in DEFAULT_RULES:
        cursor.execute("""
            INSERT INTO evaluation_rules (course_id, title, category, score_value, icon)
            VALUES (?, ?, ?, ?, ?)
        """, (course_id, rule["title"], rule["category"], rule["score_value"], rule["icon"]))
    db.commit()
    return {"message": "Default rules restored"}

@router.post("/{course_id}/add")
async def add_scores(course_id: int, data: ScoreAddRequest, db: Connection = Depends(get_db)):
    cursor = db.cursor()
    score_date = data.date or get_today_str_taipei()
    now_timestamp = get_now_str_taipei()

    # Resolve target_plan_id
    target_plan_id = data.plan_id
    if target_plan_id is None and data.group_id is not None:
        cursor.execute("SELECT plan_id FROM groups WHERE id = ?", (data.group_id,))
        grow = cursor.fetchone()
        if grow and grow["plan_id"]:
            target_plan_id = grow["plan_id"]
        else:
            cursor.execute("SELECT id FROM group_plans WHERE course_id = ? AND is_active = 1", (course_id,))
            prow = cursor.fetchone()
            if prow:
                target_plan_id = prow["id"]

    inserted_score_ids = []
    for sid in data.student_ids:
        cursor.execute("""
            INSERT INTO score_logs (course_id, student_id, rule_id, rule_title, score, category, date, timestamp, plan_id, group_id)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        """, (course_id, sid, data.rule_id, data.rule_title, data.score, data.category, score_date, now_timestamp, target_plan_id, data.group_id))
        inserted_score_ids.append(cursor.lastrowid)

    # Save Undo Log
    undo_payload = json.dumps({
        "score_log_ids": inserted_score_ids,
        "course_id": course_id,
        "score": data.score,
        "rule_title": data.rule_title,
        "count": len(data.student_ids)
    }, ensure_ascii=False)

    cursor.execute("""
        INSERT INTO undo_logs (action_type, target_id, payload_json, created_at)
        VALUES ('score', ?, ?, ?)
    """, (course_id, undo_payload, now_timestamp))
    undo_id = cursor.lastrowid

    db.commit()
    await manager.broadcast(course_id, "score_updated")
    return {
        "message": f"Successfully applied {data.score} points ({data.rule_title}) to {len(data.student_ids)} students",
        "undo_id": undo_id
    }

@router.post("/undo")
async def execute_undo(data: UndoRequest, db: Connection = Depends(get_db)):
    cursor = db.cursor()
    cursor.execute("SELECT * FROM undo_logs WHERE id = ? AND is_undone = 0", (data.undo_id,))
    undo_entry = cursor.fetchone()
    if not undo_entry:
        raise HTTPException(status_code=404, detail="Undo record not found or already undone")

    undo_entry = dict(undo_entry)
    payload = json.loads(undo_entry["payload_json"])

    if undo_entry["action_type"] == "score":
        score_log_ids = payload.get("score_log_ids", [])
        if score_log_ids:
            placeholders = ",".join(["?"] * len(score_log_ids))
            cursor.execute(f"UPDATE score_logs SET is_undone = 1 WHERE id IN ({placeholders})", score_log_ids)

    elif undo_entry["action_type"] == "attendance":
        course_id = payload["course_id"]
        target_date = payload["date"]
        old_records = payload.get("old_records", [])

        # Reset attendance records for that course & date to old state
        cursor.execute("DELETE FROM attendance WHERE course_id = ? AND date = ?", (course_id, target_date))
        for rec in old_records:
            cursor.execute("""
                INSERT INTO attendance (course_id, student_id, date, status)
                VALUES (?, ?, ?, ?)
            """, (course_id, rec["student_id"], target_date, rec["status"]))

    # Mark undo log as undone
    cursor.execute("UPDATE undo_logs SET is_undone = 1 WHERE id = ?", (data.undo_id,))
    db.commit()

    await manager.broadcast(payload["course_id"], "score_updated")
    return {"message": "Undo executed successfully"}

@router.get("/{course_id}/student/{student_id}/logs")
def get_student_score_logs(
    course_id: int,
    student_id: int,
    start_date: Optional[str] = Query(None),
    end_date: Optional[str] = Query(None),
    include_undone: bool = Query(True),
    db: Connection = Depends(get_db)
):
    cursor = db.cursor()
    query = """
        SELECT sl.id, sl.course_id, sl.student_id, sl.rule_id, sl.rule_title, sl.score, sl.category, sl.date, sl.timestamp,
               sl.is_undone, sl.plan_id, sl.group_id,
               gp.name AS plan_name,
               g.group_name
        FROM score_logs sl
        LEFT JOIN group_plans gp ON sl.plan_id = gp.id
        LEFT JOIN groups g ON sl.group_id = g.id
        WHERE sl.course_id = ? AND sl.student_id = ?
    """
    params = [course_id, student_id]
    if not include_undone or include_undone is False:
        query += " AND sl.is_undone = 0"
    if isinstance(start_date, str) and start_date:
        query += " AND sl.date >= ?"
        params.append(start_date)
    if isinstance(end_date, str) and end_date:
        query += " AND sl.date <= ?"
        params.append(end_date)
    query += " ORDER BY sl.timestamp DESC, sl.id DESC"

    cursor.execute(query, params)
    logs = [dict(row) for row in cursor.fetchall()]

    active_logs = [l for l in logs if l.get("is_undone") == 0]
    pos_sum = sum(l["score"] for l in active_logs if l["score"] > 0)
    neg_sum = sum(l["score"] for l in active_logs if l["score"] < 0)
    total_score = sum(l["score"] for l in active_logs)

    return {
        "student_id": student_id,
        "total_score": total_score,
        "positive_score": pos_sum,
        "negative_score": neg_sum,
        "logs": logs
    }

@router.delete("/{course_id}/logs/{log_id}")
async def delete_score_log(course_id: int, log_id: int, db: Connection = Depends(get_db)):
    cursor = db.cursor()
    cursor.execute("""
        UPDATE score_logs SET is_undone = 1 WHERE id = ? AND course_id = ?
    """, (log_id, course_id))
    if cursor.rowcount == 0:
        raise HTTPException(status_code=404, detail="Score log not found")
    db.commit()
    await manager.broadcast(course_id, "score_updated")
    return {"message": "Score record deleted successfully"}

@router.post("/{course_id}/logs/{log_id}/restore")
async def restore_score_log(course_id: int, log_id: int, db: Connection = Depends(get_db)):
    cursor = db.cursor()
    cursor.execute("""
        UPDATE score_logs SET is_undone = 0 WHERE id = ? AND course_id = ?
    """, (log_id, course_id))
    if cursor.rowcount == 0:
        raise HTTPException(status_code=404, detail="Score log not found")
    db.commit()
    await manager.broadcast(course_id, "score_updated")
    return {"message": "Score record restored successfully"}

File: server/src/routes/scores.ts
// Port of app/routers/scores.py
import { Router } from "express";
import { prisma } from "../db";
import { getNowStrTaipei, getTodayStrTaipei } from "../timezone";
import { broadcastToCourse } from "../realtime";
import { autoCatch } from "../asyncRoute";

export const scoresRouter = autoCatch(Router());

const DEFAULT_RULES = [
  { title: "熱心助人", category: "positive", score_value: 1, icon: "🤝" },
  { title: "發言踴躍", category: "positive", score_value: 1, icon: "🙋‍♂️" },
  { title: "專心聽講", category: "positive", score_value: 1, icon: "👂" },
  { title: "作業優良", category: "positive", score_value: 2, icon: "📝" },
  { title: "團隊合作", category: "positive", score_value: 1, icon: "🌟" },
  { title: "上課吵鬧", category: "negative", score_value: -1, icon: "📢" },
  { title: "未帶用品", category: "negative", score_value: -1, icon: "🎒" },
  { title: "上課分心", category: "negative", score_value: -1, icon: "😴" },
];

scoresRouter.get("/:courseId/rules", async (req, res) => {
  const rules = await prisma.evaluationRule.findMany({
    where: { courseId: Number(req.params.courseId) },
    orderBy: [{ category: "desc" }, { id: "asc" }],
  });
  res.json(rules);
});

scoresRouter.post("/:courseId/rules", async (req, res) => {
  const courseId = Number(req.params.courseId);
  const { title, category = "positive", score_value, icon = "⭐" } = req.body ?? {};
  const rule = await prisma.evaluationRule.create({ data: { courseId, title, category, scoreValue: score_value, icon } });
  res.json({ id: rule.id, message: "Rule created" });
});

scoresRouter.put("/:courseId/rules/:ruleId", async (req, res) => {
  const { title, category, score_value, icon } = req.body ?? {};
  await prisma.evaluationRule.updateMany({
    where: { id: Number(req.params.ruleId), courseId: Number(req.params.courseId) },
    data: { title, category, scoreValue: score_value, icon },
  });
  res.json({ message: "Rule updated" });
});

scoresRouter.post("/:courseId/rules/reset_defaults", async (req, res) => {
  const courseId = Number(req.params.courseId);
  await prisma.evaluationRule.deleteMany({ where: { courseId } });
  await prisma.evaluationRule.createMany({
    data: DEFAULT_RULES.map((r) => ({ courseId, title: r.title, category: r.category, scoreValue: r.score_value, icon: r.icon })),
  });
  res.json({ message: "Default rules restored" });
});

scoresRouter.post("/:courseId/add", async (req, res) => {
  const courseId = Number(req.params.courseId);
  const { student_ids, rule_id = null, rule_title, score, category = "positive", date, plan_id, group_id } =
    req.body ?? {};
  const scoreDate: string = date || getTodayStrTaipei();
  const nowTimestamp = getNowStrTaipei();

  let targetPlanId: number | null = plan_id ?? null;
  if (targetPlanId === null && group_id != null) {
    const g = await prisma.group.findUnique({ where: { id: group_id } });
    if (g?.planId) {
      targetPlanId = g.planId;
    } else {
      const p = await prisma.groupPlan.findFirst({ where: { courseId, isActive: 1 } });
      if (p) targetPlanId = p.id;
    }
  }

  const insertedIds: number[] = [];
  for (const sid of student_ids as number[]) {
    const log = await prisma.scoreLog.create({
      data: {
        courseId,
        studentId: sid,
        ruleId: rule_id,
        ruleTitle: rule_title,
        score,
        category,
        date: scoreDate,
        timestamp: nowTimestamp,
        planId: targetPlanId,
        groupId: group_id ?? null,
      },
    });
    insertedIds.push(log.id);
  }

  const undoPayload = JSON.stringify({
    score_log_ids: insertedIds,
    course_id: courseId,
    score,
    rule_title,
    count: (student_ids as number[]).length,
  });
  const undoLog = await prisma.undoLog.create({
    data: { actionType: "score", targetId: courseId, payloadJson: undoPayload, createdAt: nowTimestamp },
  });

  broadcastToCourse(courseId, "score_updated");
  res.json({
    message: `Successfully applied ${score} points (${rule_title}) to ${(student_ids as number[]).length} students`,
    undo_id: undoLog.id,
  });
});

scoresRouter.post("/undo", async (req, res) => {
  const undoId: number = req.body?.undo_id;
  const undoEntry = await prisma.undoLog.findFirst({ where: { id: undoId, isUndone: 0 } });
  if (!undoEntry) {
    res.status(404).json({ detail: "Undo record not found or already undone" });
    return;
  }
  const payload = JSON.parse(undoEntry.payloadJson);

  if (undoEntry.actionType === "score") {
    const ids: number[] = payload.score_log_ids ?? [];
    if (ids.length) {
      await prisma.scoreLog.updateMany({ where: { id: { in: ids } }, data: { isUndone: 1 } });
    }
  } else if (undoEntry.actionType === "attendance") {
    const courseId: number = payload.course_id;
    const targetDate: string = payload.date;
    const oldRecords: Array<{ student_id: number; status: string }> = payload.old_records ?? [];

    await prisma.attendance.deleteMany({ where: { courseId, date: targetDate } });
    for (const rec of oldRecords) {
      await prisma.attendance.create({
        data: { courseId, studentId: rec.student_id, date: targetDate, status: rec.status, updatedAt: getNowStrTaipei() },
      });
    }
  }

  await prisma.undoLog.update({ where: { id: undoId }, data: { isUndone: 1 } });
  broadcastToCourse(payload.course_id, "score_updated");
  res.json({ message: "Undo executed successfully" });
});

scoresRouter.get("/:courseId/student/:studentId/logs", async (req, res) => {
  const courseId = Number(req.params.courseId);
  const studentId = Number(req.params.studentId);
  const { start_date, end_date, include_undone = "true" } = req.query as Record<string, string>;

  const where: Record<string, unknown> = { courseId, studentId };
  if (include_undone === "false") where.isUndone = 0;
  if (start_date) where.date = { ...(where.date as object), gte: start_date };
  if (end_date) where.date = { ...(where.date as object), lte: end_date };

  const logs = await prisma.scoreLog.findMany({
    where,
    orderBy: [{ timestamp: "desc" }, { id: "desc" }],
    include: { course: false },
  });

  const planIds = [...new Set(logs.map((l) => l.planId).filter((v): v is number => v != null))];
  const groupIds = [...new Set(logs.map((l) => l.groupId).filter((v): v is number => v != null))];
  const plans = await prisma.groupPlan.findMany({ where: { id: { in: planIds } } });
  const groups = await prisma.group.findMany({ where: { id: { in: groupIds } } });
  const planNameMap = new Map(plans.map((p) => [p.id, p.name]));
  const groupNameMap = new Map(groups.map((g) => [g.id, g.groupName]));

  const enriched = logs.map((l) => ({
    id: l.id,
    course_id: l.courseId,
    student_id: l.studentId,
    rule_id: l.ruleId,
    rule_title: l.ruleTitle,
    score: l.score,
    category: l.category,
    date: l.date,
    timestamp: l.timestamp,
    is_undone: l.isUndone,
    plan_id: l.planId,
    group_id: l.groupId,
    plan_name: l.planId != null ? planNameMap.get(l.planId) ?? null : null,
    group_name: l.groupId != null ? groupNameMap.get(l.groupId) ?? null : null,
  }));

  const activeLogs = enriched.filter((l) => l.is_undone === 0);
  const posSum = activeLogs.filter((l) => l.score > 0).reduce((sum, l) => sum + l.score, 0);
  const negSum = activeLogs.filter((l) => l.score < 0).reduce((sum, l) => sum + l.score, 0);
  const totalScore = activeLogs.reduce((sum, l) => sum + l.score, 0);

  res.json({ student_id: studentId, total_score: totalScore, positive_score: posSum, negative_score: negSum, logs: enriched });
});

scoresRouter.delete("/:courseId/logs/:logId", async (req, res) => {
  const courseId = Number(req.params.courseId);
  const result = await prisma.scoreLog.updateMany({
    where: { id: Number(req.params.logId), courseId },
    data: { isUndone: 1 },
  });
  if (result.count === 0) {
    res.status(404).json({ detail: "Score log not found" });
    return;
  }
  broadcastToCourse(courseId, "score_updated");
  res.json({ message: "Score record deleted successfully" });
});

scoresRouter.post("/:courseId/logs/:logId/restore", async (req, res) => {
  const courseId = Number(req.params.courseId);
  const result = await prisma.scoreLog.updateMany({
    where: { id: Number(req.params.logId), courseId },
    data: { isUndone: 0 },
  });
  if (result.count === 0) {
    res.status(404).json({ detail: "Score log not found" });
    return;
  }
  broadcastToCourse(courseId, "score_updated");
  res.json({ message: "Score record restored successfully" });
});
